turbopack/crates/turbo-tasks/src/parallel.rs

```diff
@@ -0,0 +1,367 @@
+) -> Result<(), E>
+where
+    T: Send + Sync,
+    E: Send + 'static,
+{
+    let len = items.len();
+    if len == 0 {
+        return Ok(()); // No items to process, return early
+    }
+    let chunk_size = good_chunk_size(len);
+    let f = &f;
+    let mut result = Ok(());
+    process_in_parallel(
+        |ctx| {
+            items
+                .chunks_mut(chunk_size)
+                .map(|chunk| {
+                    ctx.task(move || {
+                        for item in chunk {
+                            f(item)?;
+                        }
+                        Ok(())
+                    })
+                })
+                .collect::<Vec<_>>()
+        },
+        |r| {
+            if let Err(e) = r {
+                if result.is_ok() {
+                    result = Err(e);
+                }
+            }
+        },
+    );
+    // SAFETY: Ensure reference is kept until here
+    let _ = items;
+    result
+}
+
+pub fn try_into_for_each<T, E>(
+    items: Vec<T>,
+    f: impl (Fn(T) -> Result<(), E>) + Send + Sync,
+) -> Result<(), E>
+where
+    T: Send + Sync,
+    E: Send + 'static,
+{
+    let len = items.len();
+    if len == 0 {
+        return Ok(()); // No items to process, return early
+    }
+    let chunk_size = good_chunk_size(len);
+    let f = &f;
+    let mut result = Ok(());
+    let mut items = unsafe { transmute::<Vec<T>, Vec<ManuallyDrop<T>>>(items) };
+    process_in_parallel(
+        |ctx| {
+            items
+                .chunks_mut(chunk_size)
+                .map(|chunk| {
+                    ctx.task(move || {
+                        for item in chunk {
+                            let item = unsafe { ManuallyDrop::take(item) };
+                            f(item)?;
+                        }
+                        Ok(())
+                    })
+                })
+                .collect::<Vec<_>>()
+        },
+        |r| {
+            if let Err(e) = r {
+                if result.is_ok() {
+                    result = Err(e);
+                }
+            }
+        },
+    );
+    // SAFETY: Ensure reference is kept until here
+    let _ = items;
+    result
+}
+
+pub fn map_collect<'l, T, I, R>(items: &'l [T], f: impl Fn(&'l T) -> I + Send + Sync) -> R
+where
+    T: Sync,
+    I: Send + Sync,
+    R: FromIterator<I>,
+{
+    let len = items.len();
+    if len == 0 {
+        return R::from_iter(std::iter::empty()); // No items to process, return empty collection
+    }
+    let chunk_size = good_chunk_size(len);
+    let f = &f;
+    let mut result = Vec::with_capacity(items.len().div_ceil(chunk_size));
+    process_in_parallel(
+        |ctx| {
+            items
+                .chunks(chunk_size)
+                .map(|chunk| {
+                    ctx.task(move || {
+                        let vec = chunk.iter().map(f).collect::<Vec<_>>();
+                        unsafe { transmute::<Vec<I>, Vec<()>>(vec) }
```

Comment:
so is this only safe because Vec<> is a glorified Box<>?  why can't we return the actual type? is `I` not Send?

Reply:
I is not 'static. I wish I could cast lifetimes of type arguments.

Reply:
Fixed that, but some tricks.